### PR TITLE
ENH: Adding vector visualization for biplots

### PR DESCRIPTION
**Pull request by mortonjt**
Adding in basic vector visualization of taxa.  This will ultimately allow easier

* Visualization of correlations between taxa
* Visualizing which taxa explains the most variability amongst the samples
* Visualizing the variance of the taxa

A couple of things will need to be addressed in the future

1. When taxa are added, both the vectors and spheres are displayed at once.  Should there be another toggle to toggle the display of the vectors?  For example, the user may want to remove the spheres if they are too clustered, but would want to keep the vectors
2. What would be a good way to color code the taxa?  They are now by default white.  It may be useful to color code each vector.  For example, when making biplots, I find it useful to color code the phyla

### CODE REVIEW

**Comment by emperor-helper** [comment]
The test build for this pull request can be found here: http://emperor.colorado.edu/pull_352/make_emperor/

File: scripts/make_emperor.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env python
-# File created on 06 Jul 2012
+# File reated on 06 Jul 2012
```
Comment (wasade) [inline scripts/make_emperor.py]: ?

Comment (mortonjt) [inline scripts/make_emperor.py]: Yup ... That was dumb ...

File: emperor/support_files/js/emperor.js [hunk 2]
```diff
@@ -735,6 +736,7 @@ function colorChanged(catValue,color) {
 function colorChangedForTaxaSpheres(color){
 	for (index in g_plotTaxa){
 		g_plotTaxa[index].material.color.setHex(color)
+
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: Can you remove this extra white-space?

File: emperor/support_files/js/emperor.js [hunk 3]
```diff
@@ -893,11 +895,14 @@ function toggleBiplotVisibility(){
 	if(!document.biplotsvisibility.elements[0].checked){
 		for (index in g_plotTaxa){
 			g_mainScene.remove(g_plotTaxa[index]);
+		        g_mainScene.remove(g_plotTaxaArrows[index]);
+
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: Can you remove this extra white-space?

File: emperor/support_files/js/emperor.js [hunk 4]
```diff
@@ -1700,6 +1715,8 @@ function makeLine(coords_a, coords_b, color, width){
 	return line;
 }
 
+
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: Can you remove this extra white-space?

**Comment by ElDeveloper** [comment]
I like this a lot, thanks!!

> When taxa are added, both the vectors and spheres are displayed at once. Should there be another toggle to toggle the display of the vectors? For example, the user may want to remove the spheres if they are too clustered, but would want to keep the vectors.

:+1: and I think we would need this as part of this PR (let me know if you need help, I'm **super** happy to Skpye/Hangouts), otherwise people who wanted to remove the vectors (and reproduce previous plots) would not be able to do it. 

> What would be a good way to color code the taxa? They are now by default white. It may be useful to color code each vector. For example, when making biplots, I find it useful to color code the phyla

This is something we have discussed a lot. There's #59, #61 and #75. If you don't think what you mean would be covered by any of this issues, please feel free to add a new issue.

------

Overall the PR looks great, I made a few minor comments and there's a few things that would be missing:

- Mention on the `ChangeLog.md`.
- Mention about these vectors in the `make_emperor.py` help.
- A way to turn them off.

File: emperor/support_files/js/emperor.js (hunk 2 above)
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: ping on this and the other extra whitespace.

File: tests/test_format.py [hunk 5]
```diff
@@ -225,26 +225,27 @@ def test_format_comparison_bars_to_js_exceptions(self):
             self.comparison_coords_headers, self.comparison_coords_data, 11)
 
     def test_format_emperor_html_footer_string(self):
+        self.maxDiff=5000
         """Test correct formatting of the footer string"""
         # footer for a jackknifed pcoa plot without biplots
         out_string = format_emperor_html_footer_string(False, True)
-        self.assertEqual(out_string, EXPECTED_FOOTER_A)
+        self.assertItemsEqual(out_string.split('/'), EXPECTED_FOOTER_A.split('/'))
```
Comment (ElDeveloper) [inline tests/test_format.py]: I'm not sure I understand what this change is doing, would you mind elaborating?

File: emperor/support_files/js/emperor.js [hunk 6]
```diff
@@ -1450,6 +1462,16 @@ function drawTaxa(){
...
 		g_mainScene.add(mesh);
 		g_plotTaxa[key] = mesh;
+
+	        // add the line from the origin to the element
+	        var taxaVector = makeLine([0, 0, 0,],
+					  [g_taxaPositions[key]['x'],
+ 					   g_taxaPositions[key]['y'],
+					   g_taxaPositions[key]['z']],
+					   whiteColor, 2);
+		g_elementsGroup.add(taxaVector)
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: These and one of the lines below is missing a semicolon at the end of the line.

File: emperor/support_files/js/emperor.js [hunk 7]
```diff
@@ -887,17 +889,27 @@ function toggleTaxaLabels(){
 }
 
 /* Turn on and off the spheres representing the biplots on screen */
-function toggleBiplotVisibility(){
+function toggleBiplotVisibility(toggleArrow){
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: Can you update the documentation string at the top?

File: emperor/support_files/js/emperor.js [hunk 8]
```diff
@@ -887,17 +889,27 @@ function toggleTaxaLabels(){
...
 	// reduce the opacity to zero if the element should be off or to 0.5
 	// if the element is supposed to be present; 0.5 is the default value
-	if(!document.biplotsvisibility.elements[0].checked){
+        toggleArrow = typeof toggleArrow !== 'undefined' ? toggleArrow : false
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: Can you add semicolons at the end of these lines?

File: emperor/format.py [hunk 9]
```diff
@@ -553,7 +553,8 @@ def format_emperor_autograph(metadata_fp, coords_fp, language='HTML'):
 _BIPLOT_VISIBILITY_SELECTOR = """
             <br>
             <form name="biplotsvisibility">
-            <input type="checkbox" onClick="toggleBiplotVisibility()" checked>Biplots Visibility</input>
+            <input type="checkbox" onClick="toggleBiplotVisibility(false)" checked>Taxa Sphere Visibility</input>
```
Comment (ElDeveloper) [inline emperor/format.py]: Can you add a `<br>` tag in-between these lines? this will make the presentation nicer:

screen shot 2015-03-04 at 2 51 44 pm

File: emperor/support_files/js/emperor.js [hunk 10]
```diff
@@ -887,17 +889,27 @@ function toggleTaxaLabels(){
...
+        arrowBox = toggleArrow ? 1 : 0
+	if(!document.biplotsvisibility.elements[arrowBox].checked){
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: I like your solution a lot!

File: ChangeLog.md [hunk 11]
```diff
@@ -13,6 +13,10 @@ Emperor 0.9.51-dev (changes since Emperor 0.9.5 go here)
 
 * Reorganized `emperor/support_files/` to separate Emperor's source code from third-party packages.
 
+### New features
+
+* Adding toggle button to display vectors from the origin to each of the taxa
```
Comment (ElDeveloper) [inline ChangeLog.md]: I would rephrase this to add more context, something like:

```
* Biplots now include arrows, which represent XXXXXXX. These can be turned off using the `Taxa arrow visibility` checkbox.

**Comment by ElDeveloper** [comment]
Looks great, I think we are only missing adding a description about these arrows somewhere in make_emperor.py and it should be ready to be merged!

File: tests/test_format.py (hunk 5 above)
Comment (mortonjt) [inline tests/test_format.py]: Its just a way to make debugging easier.  Before, an entire HTML page would be dumped on the page.  Using this will allow only portions of differing HTML to be displayed.

It isn't a big though - I can remove it.

Comment (ElDeveloper) [inline tests/test_format.py]: I see, this makes sense! A follow-up question, why did you choose '/'?
'\n' makes more sense to me, but then again, I haven't yet used this in
a real-case scenario.

Comment (mortonjt) [inline tests/test_format.py]: Just a random common letter ('/' occurs before every major html header).  I think '\n' would work just as well.

File: emperor/support_files/js/emperor.js (hunk 6 above)
Comment (mortonjt) [inline emperor/support_files/js/emperor.js]: Drats - bad python habits

Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: :snake:

File: emperor/support_files/js/emperor.js (hunk 10 above)
Comment (wasade) [inline emperor/support_files/js/emperor.js]: I think all this can be collapsed quite a bit more. My js is terrible, so I apologize if this is a bit off:

```javascript
updater = document.biplotsvisibility.elements[arrowBox].checked ? function(data, index) {g_mainScene.add(data[index])} : function(data, index) {g_mainScene.remove(data[index])};
data_to_update = toggleArrow ? g_plotTaxaArrows : g_plotTaxa;

for (index in g_plotTaxa) {
    updater(data_to_update, index);
}
```

It has the advantage of reducing the LOC and pulling out branching logic from within the forloop at the expense of some ugly lines.

File: tests/test_format.py (hunk 5 above)
Comment (ElDeveloper) [inline tests/test_format.py]: I would prefer '\n' then. Also, since assertItemsEqual doesn't care
about the order of the elements, we would need to still have the
self.assertEqual check. The first would help developers identify the
lines that need to be updated and the second verifies there's a
consistently and correctly structured HTML string.

Comment (mortonjt) [inline tests/test_format.py]: :+1:

File: emperor/support_files/js/emperor.js (hunk 10 above)
Comment (mortonjt) [inline emperor/support_files/js/emperor.js]: :+1:

**Comment by ElDeveloper** [comment]
@mortonjt, just a heads up, it seems like there's legitimate failures.

**Comment by mortonjt** [comment]
...

While I'm fixing those errors. Do you think that I should also modify the `animate`.  Namely this [line](https://github.com/biocore/emperor/blob/master/emperor/support_files/js/emperor.js#L2359)

File: emperor/support_files/js/emperor.js [hunk 12]
```diff
@@ -886,20 +888,22 @@ function toggleTaxaLabels(){
 	}
 }
 
-/* Turn on and off the spheres representing the biplots on screen */
-function toggleBiplotVisibility(){
+/* Turn on and off the spheres representing the biplots on screen
+If the toggleArrow variable is true, then only the visibility of the taxa vectors will be altered
+Otherwise, only the visibility of the */
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: Can you remove this line?

While we don't have any guidelines on this, we try to keep the comments formatted in the following way:

```javascript
/* Turn on and off the spheres representing the biplots on screen

   If the toggleArrow variable is true, then only the visibility of the taxa vectors will be altered.
*/
```

**Comment by ElDeveloper** [comment]
I think that's good as is, unless I'm missing something.

File: emperor/support_files/js/emperor.js [hunk 13]
```diff
@@ -886,20 +888,23 @@ function toggleTaxaLabels(){
 	}
 }
 
-/* Turn on and off the spheres representing the biplots on screen */
-function toggleBiplotVisibility(){
+/* Turn on and off the spheres representing the biplots on screen
+
+If the toggleArrow variable is true, then only the visibility of the taxa vectors will be altered
+Otherwise, only the visibility of the */
```
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: This line should still be removed. And the line above should be aligned with the word `Turn` that's at the beginning of this block of comments.

File: tests/test_core.py [hunk 14]
```diff
@@ -36,7 +36,7 @@ def setUp(self):
 
     def test_str(self):
         emp = Emperor(self.ord_res, self.data, self.headers)
-        self.assertEqual(HTML_STRING, str(emp))
+        self.assertItemsEqual(HTML_STRING.split('/'), str(emp).split('/'))
```
Comment (ElDeveloper) [inline tests/test_core.py]: Can you update this test case in the same way that you updated the test case in format.py?

File: emperor/support_files/js/emperor.js (hunk 2 above)
Comment (ElDeveloper) [inline emperor/support_files/js/emperor.js]: Still this extra whitespace.

**Comment by ElDeveloper** [comment]
There's just one minor (nano-comment) comment and this is ready to be merged!

File: emperor/support_files/js/emperor.js [hunk 15]
```diff
@@ -886,20 +887,23 @@ function toggleTaxaLabels(){
...
-		for (index in g_plotTaxa){
-			g_mainScene.remove(g_plotTaxa[index]);
-		}
-	}
-	else{
-		for (index in g_plotTaxa){
-			g_mainScene.add(g_plotTaxa[index])
-		}
-	}
+    toggleArrow = typeof toggleArrow !== 'undefined' ? toggleArrow : false;
```
Comment (wasade) [inline emperor/support_files/js/emperor.js]: huh, well, I guess it didn't reduce the LOC, but at least the forloop is much simpler :)

**Comment by ElDeveloper** [comment]
Thanks @mortonjt, this is ready to be merged, though there are merge conflicts, would you mind solving them?

**Comment by mortonjt** [comment]
I think that should do it

**Comment by ElDeveloper** [comment]
Thanks a lot @mortonjt!!
